flopy/modpath/mpsim.py

```diff
@@ -280,3 +280,94 @@ def write_file(self):
+        this package will be added.
+    inputstyle : 1
+        Input style described in MODPATH6 manual (currently only input style 1 is supported)
+    extension : string
+        Filename extension (default is 'loc')
+    """
+    def __init__(self, model,
+                 inputstyle=1,
+                 extension='loc',
+                 verbose=False):
+
+        Package.__init__(self, model, extension, 'LOC', 33)
+
+        self.model = model
+        self.heading = '# Starting locations file for Modpath, generated by Flopy.'
+        self.input_style = inputstyle
+        self.data = self.get_empty_starting_locations_data(0)
+        self.extension = extension
+
+        self.parent.add_package(self) # add to package list so location are written with other ModPath files
+
+    @staticmethod
+    def get_dtypes():
+        """
+           Build numpy dtype for the MODPATH 6 starting locations file.
+        """
+        dtype = np.dtype([("particleid", np.int), ("particlegroup", np.int),
+                          ('initialgrid', np.int),
+                          ('k0', np.int), ('i0', np.int),
+                          ('j0', np.int), ('xloc0', np.float32),
+                          ('yloc0', np.float32), ('zloc0', np.float32),
+                          ('initialtime', np.float32),
+                          ('label', '|S40'), ('groupname', '|S16')])
+        return dtype
+
+    @staticmethod
+    def get_empty_starting_locations_data(npt=0,
+                                          default_xloc0=0.5, default_yloc0=0.5, default_zloc0=0.):
+        """get an empty recarray for particle starting location info.
+
+        Parameters
+        ----------
+        npt : int
+            Number of particles. Particles in array will be numbered consecutively from 1 to npt.
+
+        """
+        dtype = StartingLocations.get_dtypes()
+        d = np.zeros(npt, dtype=dtype)
+        d = d.view(np.recarray)
+        d['particleid'] = np.arange(1, npt+1)
+        d['particlegroup'] = 1
+        d['initialgrid'] = 1
+        d['xloc0'] = default_xloc0
+        d['yloc0'] = default_yloc0
+        d['zloc0'] = default_zloc0
+        d['groupname'] = 'group1'
+        return d
+
+    def write_file(self, data=None, float_format='{:.8f}'):
+
+        if data is None:
+            data = self.data
+        if len(data) == 0:
+            print('No data to write!')
+            return
+        with open(self.fn_path, 'w') as output:
+            output.write('{}\n'.format(self.heading))
+            output.write('{:d}\n'.format(self.input_style))
+            groups = np.unique(self.data.groupname)
+            ngroups = len(groups)
+            output.write('{:d}\n'.format(ngroups))
+            for g in groups:
+                npt = len(self.data[self.data.groupname == g])
+                output.write('{}\n{:d}\n'.format(g.decode(), npt))
+            for p in self.data:

```

Comment:
This starts a write command per particle. This is relatively slow when you're about to simulate a lot of particles. Especially for slow hard disks. A suggestion would be to first store it in some sort of in-memory buffer. 

```python
s = ''
for p in self.data:
    s += '{:d} {:d} {:d} {:d} {:d} {:d}'.format(p.particleid, particlegroup, initialgrid, k0, p.i0, p.j0)
    fmtstr = ' {0} {0} {0} {0} '.format(float_format)
    s += fmtstr.format(p.xloc0, p.yloc0, p.zloc0, p.initialtime)
    s += '{}\n'.format(p.label.decode())

output.write(s)


```